libsignal-service/src/gv2/operations.rs

```diff
@@ -0,0 +1,221 @@
+    PendingMember as EncryptedPendingMember,
+    RequestingMember as EncryptedRequestingMember,
+};
+
+pub(crate) struct GroupOperations {
+    group_secret_params: GroupSecretParams,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum GroupDecryptionError {
+    #[error("zero-knowledge group error")]
+    ZkGroupError,
+    #[error(transparent)]
+    BincodeError(#[from] bincode::Error),
+    #[error("protobuf message decoding error: {0}")]
+    ProtobufDecodeError(#[from] prost::DecodeError),
+    #[error("wrong group attribute blob")]
+    WrongBlob,
+}
+
+impl From<zkgroup::ZkGroupError> for GroupDecryptionError {
+    fn from(_: zkgroup::ZkGroupError) -> Self {
+        GroupDecryptionError::ZkGroupError
+    }
+}
+
+impl GroupOperations {
+    fn decrypt_uuid(
+        &self,
+        uuid: &[u8],
+    ) -> Result<[u8; 16], GroupDecryptionError> {
+        let bytes = self
+            .group_secret_params
+            .decrypt_uuid(bincode::deserialize(uuid)?)
+            .map_err(|_| GroupDecryptionError::ZkGroupError)?;
+        Ok(bytes)
+    }
+
+    fn decrypt_profile_key(
+        &self,
+        profile_key: &[u8],
+        decrypted_uuid: [u8; 16],
+    ) -> Result<ProfileKey, GroupDecryptionError> {
+        Ok(self.group_secret_params.decrypt_profile_key(
+            bincode::deserialize(profile_key)?,
+            decrypted_uuid,
+        )?)
+    }
+
+    fn decrypt_member(
+        &self,
+        member: EncryptedMember,
+    ) -> Result<DecryptedMember, GroupDecryptionError> {
+        let (uuid, profile_key) = if member.presentation.is_empty() {
+            let uuid = self.decrypt_uuid(&member.user_id)?;
+            let profile_key =
+                self.decrypt_profile_key(&member.profile_key, uuid)?;
+            (uuid, profile_key)
+        } else {
+            let profile_key_credential_presentation: ProfileKeyCredentialPresentation = bincode::deserialize(&member.presentation)?;
+            let uuid = self.group_secret_params.decrypt_uuid(
+                profile_key_credential_presentation.get_uuid_ciphertext(),
+            )?;
+            let profile_key = self.group_secret_params.decrypt_profile_key(
+                profile_key_credential_presentation
+                    .get_profile_key_ciphertext(),
+                uuid,
+            )?;
+            (uuid, profile_key)
+        };
+        Ok(DecryptedMember {
+            uuid: uuid.to_vec(),
+            profile_key: bincode::serialize(&profile_key)?,
+            role: member.role,
+            joined_at_revision: member.joined_at_revision,
+        })
+    }
+
+    fn decrypt_pending_member(
+        &self,
+        member: EncryptedPendingMember,
+    ) -> Result<DecryptedPendingMember, GroupDecryptionError> {
+        let inner_member =
+            member.member.ok_or(GroupDecryptionError::WrongBlob)?;
+        let uuid = self.decrypt_uuid(&inner_member.user_id).unwrap_or_default();
```

Comment:
Maybe document why having `[0u8; _]` here is better than throwing an error for the decryption.